active/0021-rocksdb-message-persistence.md

```diff
@@ -0,0 +1,164 @@
+For the sake of simplicity we assume that `combine` function is a simple bitshift, but it should be noted that any other strategy will work mostly the same.
+Suppose every topic level takes two bits in the resulting hash and `N_l` equals 3.
+Suppose topic index of this topic filter is `101001`.
+We can create a the bitmask corresponding to this topic filter by setting bits corresponding to the wildcard levels to zeros: `110011`.
+
+Therefore iterating the database can be done using the following pseudocode:
+
+```erlang
+replay(TopicFilter) ->
+  Bitmask = gen_bitmask(TopicFilter),
+  TopicIndex = gen_topic_index(TopicFilter) band Bitmask,
+  It = db:seek(TopicIndex),
+  replay(TopicFilter, TopicIndex, Bitmask, It).
+
+replay(TopicFilter, TopicIndex, Bitmask, It) ->
+  {Key, NextIt} = db.next(It),
+  {KTopicIndex, KTimestamp} = parse_key(Key),
+  if TopicIndex bxor (KTopicIndex band Bitmask) =:= 0 ->
+       Message = db.get(Key),
+       case topic_matches(TopicFilter, Message#msg.topic) of
+         true ->
+             handle_message(Message);
+         false ->
+             %% Hash collision
+             ok
+       end,
+       replay(TopicFilter, TopicIndex, Bitmask, NextId);
+     true ->
+       %% Reached the end
+       ok
+  end.
+```
+
+Note that this schema is optimized for fast access to the topics, but it's not efficient for accessing a particular time range, since time is in the suffix of the key.
+
+There are two approaches to the problem:
+
+1. Time and topic index can be interleaved using Z-order curve, essentially organizing keys into a quadtree in `{topic_index, time}` space.
+2. Every N seconds we can create a new rocksdb column family, that holds data only for a particular time interval.
+   This approach will also make garbage collection easier, since it's possible to simply drop column families once they become outdated.
+
+However finding the best solution can only be done through experimentation.
+
+### Combining realtime messages with the historical message replay
+
+In the database schema section we proposed to sort keys by topic.
+While this is ideal for optimizing DB access while reading historical data, it leaves a significant problem:
+realtime data is not added to the DB in historical order.
+So the client replaying messages that match topics `foo/1` and `foo/2` will no longer see any messages added to `foo/1`, once it starts replaying the keys corresponding to `foo/2`.
+
+Below is the sketch of the solution (thorough testing and modelling is needed):
+
+1. Client connects with clean start =:= `false`.
+1. A client handler process is started.
+1. The client handler process initiates message replay for the topic filters from the saved session in separate processes.
+1. Once the replay reaches a next topic index (i.e. when it finishes replaying messages for index `101010` and starts replaying `101011`, it sends a message to the client handler with the last key that has `101010` prefix
```

Comment:
```suggestion
1. Once the replay reaches a next topic index (i.e. when it finishes replaying messages for index `101010` and starts replaying `101011`), it sends a message to the client handler with the last key that has `101010` prefix.
```

Reply:
If the replay is done serially by topic, won't there still be a problem of replaying the messages in historical order?